### PR TITLE
Update lua parser to be short-lived for a single filter

### PR DESCRIPTION
**Pull request by mstemm**
Clean up the interface for lua_parser/lua_parser_api so it doesn't
rely on a single object.

Context--the lua_parser object simply holds some intermediate
state (e.g. nesting level) and builds up a gen_filter as the lua side
traverses the ast of a filter expression. The lua_parser_api class
just has static methods that are registered into lua.

lua_parser used to be a single object that was reset for each
filter. Now, it's an object that is created as a single filter is
parsed and deleted afterward. The callbacks always pass the lua_parser
object as a first argument and the state/filter in the object is
updated.

Also, registering the lua callbacks is now done via a static method
instead of in the constructor.

Signed-off-by: Mark Stemm <mark.stemm@gmail.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area driver-kmod

> /area driver-ebpf

> /area libscap

/area libsinsp

> /area tests

> /area proposals

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
update: update LUA parser to be short-lived for a single filter
```

### CODE REVIEW

File: userspace/chisel/lua_parser.cpp
Comment (gnosek) [inline userspace/chisel/lua_parser.cpp:21]: 😍

File: userspace/chisel/lua_parser.cpp
Comment (gnosek) [inline userspace/chisel/lua_parser.cpp:45]: This can't be done in the initializer list I expect?

File: userspace/chisel/lua_parser.cpp
Comment (gnosek) [inline userspace/chisel/lua_parser.cpp:48]: Is there still a reason for the destructor to exist?

File: userspace/chisel/lua_parser.cpp
Comment (gnosek) [inline userspace/chisel/lua_parser.cpp:66]: Can we return `std::shared_ptr<T>&` and maybe save a refcount inc/dec? No idea whether it's worth it so your call

File: userspace/chisel/lua_parser.cpp
Comment (gnosek) [inline userspace/chisel/lua_parser.cpp:55]: I'm only worried about how expensive that call is (now that we call it more often... I guess?)

**Review by gnosek — COMMENTED** [review]
-ENOCONTEXT but looks okay at a glance

File: userspace/chisel/lua_parser.cpp
Comment (mstemm) [reply userspace/chisel/lua_parser.cpp:21]: I knew you would like it!

File: userspace/chisel/lua_parser.cpp
Comment (mstemm) [reply userspace/chisel/lua_parser.cpp:45]: No, it can. I'll change it.

File: userspace/chisel/lua_parser.cpp
Comment (mstemm) [reply userspace/chisel/lua_parser.cpp:48]: Just completeness, I always define a destructor even if it doesn't do anything.

File: userspace/chisel/lua_parser.cpp
Comment (mstemm) [reply userspace/chisel/lua_parser.cpp:66]: I prefer to not play games like that for shared_ptrs and just let the refcounting do it's thing. This isn't performance-sensitive really, as it only happens when compiling filters, which is relatively uncommon.

File: userspace/chisel/lua_parser.cpp
Comment (mstemm) [reply userspace/chisel/lua_parser.cpp:55]: No, the method is static and it's only called once when the falco engine is created.

**Review by leogr — APPROVED** [review]
LGTM
/approve


/hold
For the reasons explained [here](https://github.com/falcosecurity/libs/pull/75#pullrequestreview-756411175)

**Review by gnosek — APPROVED** [review]

**Comment by mstemm** [comment]
Falco 0.30.0 is released now, so I'm cancelling the hold.
/hold cancel

**Pull request merged**
